base/bucket_gocb.go

```diff
@@ -853,17 +853,18 @@ func (bucket CouchbaseBucketGoCB) WriteCas(k string, flags int, exp int, cas uin
 // CAS-safe write of a document and it's associated named xattr
 func (bucket CouchbaseBucketGoCB) WriteCasWithXattr(k string, xattrKey string, exp int, cas uint64, v interface{}, xv interface{}) (casOut uint64, err error) {
 
+	// WriteCasWithXattr always stamps the xattr with the new cas using macro expansion, into a top-level property called 'cas'.
+	// This is the only use case for macro expansion today - if more cases turn up, should change the bucket API to handle this more generically.
+	xattrCasProperty := fmt.Sprintf("%s.cas", xattrKey)
+
 	// If cas=0, treat as insert
 	if cas == 0 {
-		// TODO: do this insert and update in a single op, when https://issues.couchbase.com/browse/MB-23522 is fixed.
 		// TODO: Once that's fixed, need to wrap w/ retry handling
 		gocbExpvars.Add("WriteCasWithXattr_Insert", 1)
-		newCas, err := bucket.Bucket.Insert(k, &v, uint32(exp))
-		if err != nil {
-			return 0, err
-		}
-		docFragment, err := bucket.Bucket.MutateIn(k, newCas, uint32(exp)).
-			UpsertEx(xattrKey, xv, gocb.SubdocFlagXattr|gocb.SubdocFlagCreatePath).
+		docFragment, err := bucket.Bucket.MutateInEx(k, gocb.SubdocDocFlagReplaceDoc, 0, uint32(exp)).
+			UpsertEx(xattrKey, xv, gocb.SubdocFlagXattr).
+			UpsertEx(xattrCasProperty, "${Mutation.CAS}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros).
+			UpsertEx("", v, gocb.SubdocFlagNone).
```

Comment:
Why is this needed?  `UpsertEx("", v, gocb.SubdocFlagNone).`  The `gocb.SubdocFlagNone` docs say "indicate no special behaviors", but in previous method call it's passing `gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros`.  Also what does it mean to pass `""` in path?